Stronger type checking and other WAL improvements

ctubbsii opened the pull request:
* Use LogEntry type in more places where just the logEntry's filename as a String was being used
* Keep the UUID and HostAndPort in LogEntry that are created for validation, rather than their String equivalents
* Remove no longer needed deleteWal(String) and create a deleteFromMutation method on LogEntry to keep the schema serialization/deserialization in LogEntry (may want to move that to LogColumnFamily class in the MetadataSchema later, but for now, at least it's in one place)
* Use private constructor in LogEntry, do validation in a fromFilePath static method before calling the constructor with the strongly typed results
* Update javadocs to clarify the filename format
* Fix the validation of the host+port in the filename (it's a plus sign not a colon, because colons weren't valid characters in HDFS, so we used a plus instead)
* Simplify VolumeUtil switchVolumes logic that was previously complicated because an old LogEntry format could contain multiple paths per entry and that's no longer the case
* Clean up DfsLogger a little bit to remove some old stuff that wasn't needed: remove ServerResources used for testing when a mock ServerContext would suffice, pass LogEntry instead of Strings for the filename and its metadata format (the meta reference being passed wasn't actually what was in the table anyway, it was coming from LogEntry, which was computing it, and Ample wasn't using this meta for deletions/cleanup anyway - may be an existing bug; if so, that will be a follow-on), stop using toString() to compute serialization and remove the redundant use of the logger name itself to try to compute the serialization in that toString() method

DomGarguilo commented on core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java:
```diff
@@ -27,64 +27,106 @@
    * @throws IllegalArgumentException if the filePath is invalid
    */
-  private static void validateFilePath(String filePath) {
+  public static LogEntry fromFilePath(String filePath) {
```
Seems fine without it but do you think it would be helpful to have another static entry point that accepts a `Path` object? Or even replace the type of the `filePath` member variable with `Path`?

It seems like a lot of the places in the code convert from a `Path` to a `String` when using this method.

ctubbsii replied:
I thought of that, but I didn't for a few reasons:
1. I'm hesitant to more tightly couple HDFS types in code when I'd like to eventually get all of it out of our code (as much as possible)
2. There's already some confusion in that the Hadoop Path type is used in two of the three relevant places in code, but the Java NIO Path type is used in tests, and toString works for both right now, so I didn't want to make the changes too big
3. I'm uncertain the impact of using new Path for the HDFS Path type, if there's any kind of normalization or transformation, and I wanted to do simple refactorings, without changing any behavior that might risk breakage

DomGarguilo reviewed (commented):
Haven't reviewed everything yet but what I have reviewed looks good so far.

EdColeman commented on server/tserver/src/main/java/org/apache/accumulo/tserver/TabletClientHandler.java:
```diff
@@ -1281,12 +1282,9 @@ public void compactionJobFailed(TInfo tinfo, TCredentials credentials,
 
   @Override
   public List<String> getActiveLogs(TInfo tinfo, TCredentials credentials) {
```
As a follow-on, could this be a List<LogEntry> instead of strings?

ctubbsii replied:
That would require making LogEntry a Thrift type, as this method is implementing a Thrift interface. I think it's fine to serialize it as a String for the network API.

EdColeman reviewed (approved):
LGTM.   The may be somethings that can be looked as a follow-on, but this really cleans things up and helps future changes stand-out and not get lost in a big PR.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java:
```diff
@@ -27,64 +27,106 @@
-      throw new IllegalArgumentException("Expected valid UUID. Found '" + uuidPart + "'");
+      throw new IllegalArgumentException(badUuidMsg);
     }
+    if (!uuid.toString().equals(uuidPart)) {
```
What error case does this check cover that UUID.fromString does not?

EdColeman replied:
UUID.fromString accepts 1-1-1-1-1

ctubbsii replied:
`UUID.fromString("1-1-1-2-3")` will parse as UUID `00000001-0001-0001-0002-000000000003`, even though it's not a valid UUID format that we expect.

keith-turner commented on core/src/test/java/org/apache/accumulo/core/tabletserver/log/LogEntryTest.java:
```diff
@@ -25,56 +25,60 @@
+  private final String validHost = "localhost+9997";
   private final UUID validUUID = UUID.randomUUID();
-  private final String validFilename = validHost + "/" + validUUID;
+  private final String validFilename = "viewfs:/a/accumulo/wal/" + validHost + "/" + validUUID;
```
could rename this variable to validPath or validUri

keith-turner commented on core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java:
```diff
@@ -27,64 +27,106 @@
 
+    String badTServerMsg =
+        "Invalid tserver in filePath. Expected: host+port. Found '" + tserverPart + "'";
+    if (tserverPart.contains(":")) {
```
Would it make sense to also check for `+`?

```suggestion
    if (tserverPart.contains(":") || !tseverPart.contains("+")) {
```

ctubbsii replied:
I wasn't sure if port was required, but I think it is, so I'll add this.

keith-turner reviewed (approved):
This is a nice cleanup.

ctubbsii commented:
I think I addressed all the comments. I have a follow-up to this in PR #4085 that also addresses issue #4061, and created issue #4086 as another follow-on, that is only loosely related.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/tabletserver/log/LogEntry.java:
```diff
@@ -27,103 +27,153 @@
     String uuidPart = parts[parts.length - 1];
 
+    String badTServerMsg =
+        "Invalid tserver in path. Expected: host+port. Found '" + tserverPart + "'";
+    if (tserverPart.contains(":") || !tserverPart.contains("+")) {
+      throw new IllegalArgumentException(badTServerMsg);
+    }
+    HostAndPort tserver;
     try {
-      HostAndPort.fromString(tserverPart);
+      tserver = HostAndPort.fromString(tserverPart.replace("+", ":"));
```
You can't use server.toString(), because it produces a string with a colon in it, and that's [not valid for HDFS](https://hadoop.apache.org/docs/r3.3.6/hadoop-project-dist/hadoop-common/filesystem/model.html).

When we write this to a file, we replace the colons with a plus sign, and vice versa. That may not have been obvious in your tests because you used java.io.file.Path, rather than Hadoop's Path type.

If you don't convert the plus sign to a colon, HostAndPort will still parse it, but it will put everything in the host portion, and leave the port empty. So:

```java
HostAndPort.fromString("localhost+8080"); // results in {host=localhost+8080, port=}
HostAndPort.fromString("localhost:8080"); // results in {host=localhost, port=8080}
```
It's hard to tell, because HostAndPort.toString() looks the same in both cases.

When we write the file to disk, we actually convert it to use the plus sign:

https://github.com/apache/accumulo/blob/f9897862dd4e6ff4892239ff5ebeb8ed6e34bc68/server/tserver/src/main/java/org/apache/accumulo/tserver/log/DfsLogger.java#L384
